pkg/controller/argocd/keycloak.go

```diff
@@ -0,0 +1,385 @@
+				"service.alpha.openshift.io/serving-cert-secret-name":    "sso-x509-jgroups-secret",
+			},
+		},
+		TypeMeta: metav1.TypeMeta{APIVersion: "v1", Kind: "Service"},
+		Spec: corev1.ServiceSpec{
+			ClusterIP: "None",
+			Ports: []corev1.ServicePort{
+				{Name: "ping", Port: 8888},
+			},
+			Selector: map[string]string{
+				"deploymentConfig": "${APPLICATION_NAME}",
+			},
+		},
+	}
+}
+
+func getKeycloakRouteTemplate(ns string) *routev1.Route {
+	return &routev1.Route{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:      map[string]string{"application": "${APPLICATION_NAME}"},
+			Name:        "${APPLICATION_NAME}",
+			Namespace:   ns,
+			Annotations: map[string]string{"description": "Route for application's https service"},
+		},
+		TypeMeta: metav1.TypeMeta{APIVersion: "v1", Kind: "Route"},
+		Spec: routev1.RouteSpec{
+			TLS: &routev1.TLSConfig{
+				Termination: "reencrypt",
+			},
+			To: routev1.RouteTargetReference{
+				Name: "${APPLICATION_NAME}",
+			},
+		},
+	}
+}
+
+func newKeycloakTemplateInstance(ns string) (*template.TemplateInstance, error) {
+	tpl, err := newKeycloakTemplate(ns)
+	if err != nil {
+		return &template.TemplateInstance{}, err
+	}
+	return &template.TemplateInstance{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "rhsso",
+			Namespace: ns,
+		},
+		Spec: template.TemplateInstanceSpec{
+			Template: tpl,
+		},
+	}, nil
+}
+
+func newKeycloakTemplate(ns string) (template.Template, error) {
+	tmpl := template.Template{}
+	configMapTemplate := getKeycloakConfigMapTemplate(ns)
+	deploymentConfigTemplate := getKeycloakDeploymentConfigTemplate(ns)
+	serviceTemplate := getKeycloakServiceTemplate(ns)
+	pingServiceTemplate := getKeycloakPingServiceTemplate(ns)
+	routeTemplate := getKeycloakRouteTemplate(ns)
+
+	configMap, err := json.Marshal(configMapTemplate)
+	if err != nil {
+		return tmpl, err
+	}
+
+	deploymentConfig, err := json.Marshal(deploymentConfigTemplate)
+	if err != nil {
+		return tmpl, err
+	}
+
+	service, err := json.Marshal(serviceTemplate)
+	if err != nil {
+		return tmpl, err
+	}
+
+	pingService, err := json.Marshal(pingServiceTemplate)
+	if err != nil {
+		return tmpl, err
+	}
+
+	route, err := json.Marshal(routeTemplate)
+	if err != nil {
+		return tmpl, err
+	}
+
+	tmpl = template.Template{
+		ObjectMeta: metav1.ObjectMeta{
+			Annotations: map[string]string{
+				"description":               "RH-SSO Template for Gitops Operator",
+				"iconClass":                 "icon-sso",
+				"openshift.io/display-name": "Keycloak",
+				"tags":                      "keycloak",
+				"version":                   "9.0.4-SNAPSHOT",
+			},
+			Name:      "rhsso",
+			Namespace: ns,
+		},
+		Objects: []runtime.RawExtension{
+			{
+				Raw: json.RawMessage(configMap),
+			},
+			{
+				Raw: json.RawMessage(deploymentConfig),
+			},
+			{
+				Raw: json.RawMessage(service),
+			},
+			{
+				Raw: json.RawMessage(pingService),
+			},
+			{
+				Raw: json.RawMessage(route),
+			},
+		},
```

Comment:
After reading these code, I am not sure if we really need template. :-)   What we are doing here is to create a configmap, deploymentConfig/deployment (you don't need deploymentConfig unless we want re-deploy if image changes), keycloak service, ping service, and a route.

Reply:
I  raised this question in the last design meeting. I understood that we need this as this is supported way of deployment and support team is comfortable with this.

Also, deployment config comes with image streams which is an added advantage.